File: Objective-C/Common/BaseViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A base table view controller to share a data model between subclasses. Allows its subclasses to display product and purchase information.
*/


#import "Section.h"
#import "BaseViewController.h"

@implementation BaseViewController

- (instancetype)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];

	if (self != nil) {
		_data = [[NSMutableArray alloc] initWithCapacity:0];
	}
	return self;
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
	// Returns the number of sections.
	return self.data.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	Section *model = (Section *)self.data[section];
	// Returns the number of rows in the section.
	return model.elements.count;
}

-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
	Section *model = (Section *)self.data[section];
	// Returns the header title for this section.
	return model.name;
}

#pragma mark - IAPTableViewDataSource

-(void)reloadWithData:(NSArray *)data {
	self.data = [data copy];
	[self.tableView reloadData];
}

@end

File: Swift/Common/BaseViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A base table view controller to share a data model between subclasses. Allows its subclasses to display product and purchase information.
*/

import UIKit

class BaseViewController: UITableViewController {
	// MARK: - Properties

	/// Data model used by all BaseViewController subclasses.
	var data = [Section]()

	// MARK: - UITableViewDataSource

	override func numberOfSections(in tableView: UITableView) -> Int {
		// Returns the number of sections.
		return data.count
	}

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		// Returns the number of rows in the section.
		return data[section].elements.count
	}

	override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		// Returns the header title for this section.
		return data[section].type.description
	}
}

// MARK: - BaseViewController Extension

/// Extends BaseViewController to refresh the UI with new data.
extension BaseViewController {
	func reload(with data: [Section]) {
		self.data = data
		tableView.reloadData()
	}
}
